src/Microsoft.AspNetCore.Mvc.Razor/Internal/RazorPagePropertyActivator.cs

```diff
@@ -0,0 +1,214 @@
+        private Func<ViewContext, object> _urlHelperAccessor;
+        private Func<ViewContext, object> _jsonHelperAccessor;
+        private Func<ViewContext, object> _diagnosticSourceAccessor;
+        private Func<ViewContext, object> _htmlEncoderAccessor;
+        private Func<ViewContext, object> _modelExpressionProviderAccessor;
+
+        public RazorPagePropertyActivator(
+            IModelMetadataProvider metadataProvider,
+            IUrlHelperFactory urlHelperFactory,
+            IJsonHelper jsonHelper,
+            DiagnosticSource diagnosticSource,
+            HtmlEncoder htmlEncoder,
+            IModelExpressionProvider modelExpressionProvider)
+        {
+            _activationInfo = new ConcurrentDictionary<Type, PageActivationInfo>();
+            _metadataProvider = metadataProvider;
+            _urlHelperAccessor = context => urlHelperFactory.GetUrlHelper(context);
+            _jsonHelperAccessor = context => jsonHelper;
+            _diagnosticSourceAccessor = context => diagnosticSource;
+            _htmlEncoderAccessor = context => htmlEncoder;
+            _modelExpressionProviderAccessor = context => modelExpressionProvider;
+        }
+
+        public void Activate(IRazorPage page, ViewContext context, Type modelType)
+        {
+            if (page == null)
+            {
+                throw new ArgumentNullException(nameof(page));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            PageActivationInfo activationInfo;
+            if (!_activationInfo.TryGetValue(page.GetType(), out activationInfo))
+            {
+                activationInfo = _activationInfo.GetOrAdd(
+                    page.GetType(),
+                    CreateViewActivationInfo(page.GetType(), modelType));
+            }
+
+            context.ViewData = CreateViewDataDictionary(context, activationInfo);
+
+            for (var i = 0; i < activationInfo.PropertyActivators.Length; i++)
+            {
+                var activateInfo = activationInfo.PropertyActivators[i];
+                activateInfo.Activate(page, context);
+            }
+        }
+
+        private ViewDataDictionary CreateViewDataDictionary(ViewContext context, PageActivationInfo activationInfo)
+        {
+            // Create a ViewDataDictionary<TModel> if the ViewContext.ViewData is not set or the type of
+            // ViewContext.ViewData is an incompatible type.
+            if (context.ViewData == null)
+            {
+                // Create ViewDataDictionary<TModel>(IModelMetadataProvider, ModelStateDictionary).
+                return activationInfo.CreateViewDataRoot(
+                    _metadataProvider,
+                    context.ModelState);
+            }
+            else if (context.ViewData.GetType() != activationInfo.ViewDataDictionaryType)
```

Comment:
Ah, I see, you're trying to de-dupe code. Could we extract just the viewdata part to its own class?